File: DSA Practice/LeetCode/2130. Maximum Twin Sum of a Linked List/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int pairSum(ListNode head) {
        Stack<Integer> stack = new Stack<Integer>();
        ListNode ref = head;
        int length = 0;
        while(head != null){
            stack.push(head.val);
            length++;
            head = head.next;
        }

        int ma = Integer.MIN_VALUE;
        for (int i = 0; i <= (length/2) - 1; i++){
            int n1 = stack.pop();
            int n2 = ref.val;

            ma = Math.max(ma, n1+n2);

            ref = ref.next;
        }

        return ma;
    }
}

File: DSA Practice/LeetCode/2130. Maximum Twin Sum of a Linked List/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def pairSum(self, head: Optional[ListNode]) -> int:
        ref = head
        length = 0
        stack = []

        while (head):
            stack.append(head.val)
            length += 1
            head = head.next

        ma = -1

        for i in range(length//2):
            n1 = stack.pop()
            n2 = ref.val

            ma = max(ma, n1 + n2)

            ref = ref.next


        return ma